src/validation.cpp

```diff
@@ -914,7 +914,12 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
 
     // No transactions are allowed below minRelayTxFee except from disconnected
     // blocks
-    if (!bypass_limits && !CheckFeeRate(ws.m_vsize, ws.m_modified_fees, state)) return false;
+    bool fee_check = CheckFeeRate(ws.m_vsize, ws.m_modified_fees, state);
```

Comment:
This is true, and it may be clearer to make this an `if`/`else`, but the poisoning of `state` doesn't matter the way that it's used in this codebase. `state` is only accessed when a function returns a falsey value (which are all of the form `state.Invalid` which re-poisons `state` overwriting any previous incorrect value). So either everything else succeeds after this line, and `state` is never accessed, or something else false, and `state` is correctly set to `Invalid`.

But great catch! Let's fix this to prevent any future bugs.

This business of using pseudo-globals to emulate monadic error handling is peak C++..

Reply:
fixed in e6b0945f0d3c4c1228d27fcdd8cca205f95f0f89

Reply:
You applied to the `return state.Valid();` to the wrong function.

Furthermore, IMHO, we want to endevour to minimize the diff between upstream bitcoin and elements in order to make merging as easy as we can reasonably get away with.  Granted, I'm not in charge of elements development, but adding a whole new method to `class ValidationState` doesn't strike me as minimizing the diff, especially when all we need to do is use an if statement or a conditional operator.

e.g. just do the following:

```
   if (!bypass_limits && !CheckFeeRate(Params().GetAcceptDiscountCT() ? GetDiscountVirtualTransactionSize(tx) : ws.m_vsize, ws.m_modified_fees, state)) return false;
```
or, if you think that line is too long (which it probably is) then

```
   int64_t fee_rate_tx_size = Params().GetAcceptDiscountCT() ? GetDiscountVirtualTransactionSize(tx) : ws.m_vsize
   if (!bypass_limits && !CheckFeeRate(fee_rate_tx_size, ws.m_modified_fees, state)) return false;
```

This keeps the original (contrived) logic that Bitcoin uses for error handling in place.

Reply:
Thanks, reverted that commit and updated in 70b9c024b158a45e7aeade64033382183d803bf5